mgmtfn/k8splugin/driver.go

```diff
@@ -0,0 +1,371 @@
+			cidr, newname, err)
+		return nil
+	}
+	log.Infof("Output from ip assign: %v", assignIP)
+
+	// Finally, mark the link up
+	bringUp, err := osexec.Command(nsenterPath, "-t", nsPid, "-n", "-F", "--", ipPath,
+		"link", "set", "dev", newname, "up").CombinedOutput()
+
+	if err != nil {
+		log.Errorf("unable to assign ip %s to %s. Error: %s",
+			cidr, newname, err)
+		return nil
+	}
+	log.Infof("Output from ip assign: %v", bringUp)
+	return nil
+
+}
+
+// getEPSpec gets the EP spec using the pod attributes
+func getEPSpec(pInfo *cniapi.CNIPodAttr) (*epSpec, error) {
+	resp := epSpec{}
+
+	// Get labels from the kube api server
+	epg, err := kubeAPIClient.GetPodLabel(pInfo.K8sNameSpace, pInfo.Name, "net-group")
+	if err != nil {
+		log.Errorf("Error getting epg. Err: %v", err)
+		return &resp, err
+	}
+
+	// Safe to ignore the error return for subsequent invocations of GetPodLabel
+	netw, _ := kubeAPIClient.GetPodLabel(pInfo.K8sNameSpace, pInfo.Name, "network")
+	tenant, _ := kubeAPIClient.GetPodLabel(pInfo.K8sNameSpace, pInfo.Name, "tenant")
+	log.Infof("labels is %s/%s/%s for pod %s\n", tenant, netw, epg, pInfo.Name)
+	resp.Tenant = tenant
+	resp.Network = netw
+	resp.Group = epg
+	resp.EndpointID = pInfo.InfraContainerID
+
+	return &resp, nil
+}
+
+// addPod is the handler for pod additions
+func addPod(r *http.Request) (interface{}, error) {
+
+	resp := cniapi.RspAddPod{}
+
+	logEvent("add pod")
+
+	content, err := ioutil.ReadAll(r.Body)
+	if err != nil {
+		log.Errorf("Failed to read request: %v", err)
+		return resp, err
+	}
+
+	pInfo := cniapi.CNIPodAttr{}
+	if err := json.Unmarshal(content, &pInfo); err != nil {
+		return resp, err
+	}
+
+	// Get labels from the kube api server
+	epReq, err := getEPSpec(&pInfo)
+	if err != nil {
+		log.Errorf("Error getting labels. Err: %v", err)
+		return resp, err
+	}
+
+	ep, err := createEP(epReq)
+	if err != nil {
+		log.Errorf("Error creating ep. Err: %v", err)
+		return resp, err
+	}
+
+	// Set interface attributes for the new port
+	err = setIfAttrs(ep.PortName, pInfo.NwNameSpace, ep.IPAddress, pInfo.IntfName)
+	if err != nil {
+		log.Errorf("Error setting interface attributes. Err: %v", err)
+		return resp, err
+	}
+
+	resp.IPAddress = ep.IPAddress
+	resp.EndpointID = pInfo.InfraContainerID
+	return resp, nil
+}
+
+// deletePod is the handler for pod deletes
+func deletePod(r *http.Request) (interface{}, error) {
+
+	resp := cniapi.RspAddPod{}
+
+	logEvent("del pod")
+
+	content, err := ioutil.ReadAll(r.Body)
+	if err != nil {
+		log.Errorf("Failed to read request: %v", err)
+		return resp, err
+	}
+
+	pInfo := cniapi.CNIPodAttr{}
+	if err := json.Unmarshal(content, &pInfo); err != nil {
+		return resp, err
+	}
+
+	// Get labels from the kube api server
+	epReq, err := getEPSpec(&pInfo)
+	if err != nil {
+		log.Errorf("Error getting labels. Err: %v", err)
+		return resp, err
+	}
+
+	err = epCleanUp(epReq)
+	resp.EndpointID = pInfo.InfraContainerID
+	return resp, err
+}
```

Comment:
I have added UT for the REST interface. This one, seems like it will need to be covered by system test as this module makes multiple external calls.